sdk/search/azure-search-documents/azure/search/documents/_internal/_search_index_document_batching_client.py

```diff
@@ -0,0 +1,280 @@
+
+import threading
+
+from azure.core.tracing.decorator import distributed_trace
+from azure.core.exceptions import HttpResponseError
+from .._api_versions import validate_api_version
+from ._utils import is_retryable_status_code
+from ._generated import SearchIndexClient
+from ._generated_serviceclient import SearchServiceClient
+from ._generated.models import IndexBatch, IndexingResult
+from ._search_documents_error import RequestEntityTooLargeError
+from ._index_documents_batch import IndexDocumentsBatch
+from .._headers_mixin import HeadersMixin
+from .._version import SDK_MONIKER
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Union
+    from azure.core.credentials import AzureKeyCredential
+
+class SearchIndexDocumentBatchingClient(HeadersMixin):
+    """A client to do index document batching.
+
+    :param endpoint: The URL endpoint of an Azure search service
+    :type endpoint: str
+    :param index_name: The name of the index to connect to
+    :type index_name: str
+    :param credential: A credential to authorize search client requests
+    :type credential: ~azure.core.credentials.AzureKeyCredential
+    :keyword int window: how many seconds if there is no changes that triggers auto flush.
+        if window is less or equal than 0, it will disable auto flush
+    :keyword int batch_size: batch size. It only takes affect when auto_flush is on
+    :keyword persistence: persistence hook. If it is set, the batch client will dump actions queue when it changes
+    :keyword str api_version: The Search API version to use for requests.
+    """
+    # pylint: disable=too-many-instance-attributes
+    _ODATA_ACCEPT = "application/json;odata.metadata=none"  # type: str
+    _DEFAULT_WINDOW = 0
+    _DEFAULT_BATCH_SIZE = 1000
+
+    def __init__(self, endpoint, index_name, credential, **kwargs):
+        # type: (str, str, AzureKeyCredential, **Any) -> None
+
+        api_version = kwargs.pop('api_version', None)
+        validate_api_version(api_version)
+        self._batch_size = kwargs.pop('batch_size', self._DEFAULT_BATCH_SIZE)
+        self._window = kwargs.pop('window', self._DEFAULT_WINDOW)
+        self._auto_flush = self._window > 0
+        self._index_documents_batch = IndexDocumentsBatch()
+        self._endpoint = endpoint  # type: str
+        self._index_name = index_name  # type: str
+        self._index_key = None
+        self._credential = credential  # type: AzureKeyCredential
+        self._client = SearchIndexClient(
+            endpoint=endpoint, index_name=index_name, sdk_moniker=SDK_MONIKER, **kwargs
+        )  # type: SearchIndexClient
+        self._timer = threading.Timer(self._window, self.flush)
+        if self._auto_flush:
+            self._timer.start()
+        self._persistence = kwargs.pop('persistence', None)
+
+    def cleanup(self):
+        # type: () -> None
+        if self._auto_flush:
+            self._timer.cancel()
```

Comment:
Should clean-up flush? Is there a risk to clean-up and lose stuff unflushed? Regular protocol like files IO will flush on close.